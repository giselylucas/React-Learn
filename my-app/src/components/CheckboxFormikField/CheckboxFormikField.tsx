import React from 'react';
import { Field, FieldProps } from 'formik';

interface Props{
    name: string;
    label: string;
    className: string;
}

function CheckboxFormikField(props: Props) {
    return (
        <Field
            name={props.name}
            className={props.className}>
        {({field}: FieldProps) => (
            <div>
                <label htmlFor="hasCorn1">
                    {props.label}
                </label>
                <input
                    id="hasCorn1"
                    type="checkbox"
                    {...field}
                />
            </div>
        )}
        </Field>
    );
}

export default CheckboxFormikField;