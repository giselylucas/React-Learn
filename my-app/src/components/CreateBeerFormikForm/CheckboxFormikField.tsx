import React from 'react';
import { Field, FieldProps } from 'formik';

interface Props{
    name: string;
    label: string;
    className: string;
}

function CheckboxFormikField(props: Props) {
    return (
        <Field
            name={props.name}
            className={props.className}>
        {({field, form}: FieldProps) => (
            <label>
            <input
                type="checkbox"
                {...field}
            />
            {props.label}
            </label>
        )}
        </Field>
    );
}

export default CheckboxFormikField;