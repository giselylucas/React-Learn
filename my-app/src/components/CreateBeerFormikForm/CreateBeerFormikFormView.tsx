import React from 'react';
import { Formik, Form, Field, ErrorMessage, FieldProps, FormikProps } from 'formik';
import { Schema } from 'yup';
import Checkbox from './CheckboxFormikField';

interface Props<T> {
    initialValues: T;
    onSubmit: (values: T) => void;
    validationSchema: Schema<T>;
}

function CreateBeerFormikFormView (props: Props<T>) {
    return (
        <Formik
                initialValues={props.initialValues}
                onSubmit={props.onSubmit}
                validationSchema={props.validationSchema}>
                {(props: FormikProps<any>) => (
                    <Form onSubmit={props.handleSubmit}>
                        <label htmlFor="beerName1">Beer Name</label>
                        <Field
                            className="newBeerName"
                            id="beerName1"
                            name="beerName"/>
                        <ErrorMessage name="beerName" >
                            {errorMessage => <p>{errorMessage}</p>}
                        </ErrorMessage>
                        <label htmlFor="beerType1">Beer Type</label>
                        <Field 
                            id="beerType1"
                            className="newBeerType"
                            component="select"
                            name="beerType" 
                            defaultValue="">
                            <option
                                value=""
                                hidden >
                                Select a option
                            </option>
                            <option value="ale">Ale</option>
                            <option value="lager">Lager</option>
                            <option value="stout">Stout</option>
                        </Field>
                        <ErrorMessage name="beerType" >
                            {errorMessage => <p>{errorMessage}</p>}
                        </ErrorMessage>
                        <Checkbox
                            className="newHasCorn"
                            name="hasCorn"
                            label="Has Corn"/>
                        <label htmlFor="ingredients1">Ingredients</label>
                        <Field
                            id="ingredients1"
                            className="newIngredients"
                            component="textarea"
                            name="ingredients" />
                        <ErrorMessage name="ingredients" >
                            {errorMessage => <p>{errorMessage}</p>}
                        </ErrorMessage>
                        <button
                            type="submit"
                            disabled={!(props.isValid && props.dirty)}>
                            Submit
                        </button>
                    </Form>
                )}
            </Formik>
    );
};

export default CreateBeerFormikFormView;