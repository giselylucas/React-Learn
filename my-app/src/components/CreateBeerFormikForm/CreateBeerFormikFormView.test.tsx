import React from 'react';
import { shallow } from 'enzyme';

import CreateBeerFormikFormView from './CreateBeerFormikFormView';
import * as Yup from 'yup';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import CheckboxFormikField from '../CheckboxFormikField/CheckboxFormikField';

describe('CreateBeerFormikFormView', () => {
    const props = {
        initialValues: {},
        validationSchema: Yup.object().shape({}),
        onSubmit: jest.fn()
    };

    const formikProps = {
        handleSubmit: jest.fn(),
        isValid: false,
        isDirty: true
    };

    const errorMessage = (errorId: String) => {
        return <ErrorMessage name={errorId}></ErrorMessage>;
    };

    const wrapper = shallow(<CreateBeerFormikFormView 
                                initialValues={props.initialValues}
                                validationSchema={props.validationSchema}
                                onSubmit={props.onSubmit}
                            />);

    describe('Render the component main structure', () => {
        it('should to render this component', () => {        
            expect(wrapper).toHaveLength(1);
    
            const formikWrapper = wrapper.find(Formik);
    
            expect(formikWrapper).toHaveLength(1);
            expect(formikWrapper.prop('initialValues')).toBe(props.initialValues);
            expect(formikWrapper.prop('onSubmit')).toBe(props.onSubmit);
            expect(formikWrapper.prop('validationSchema')).toBe(props.validationSchema);
    
        });
    
        it('should to render correctly children', () => {
            const formikWrapper = wrapper
                                    .find(Formik)
                                    .renderProp('children')(formikProps);
    
            expect(formikWrapper.find(Form)).toHaveLength(1);
            expect(formikWrapper.children()).toHaveLength(11);
        });
        
    });

    describe('Render correctly order of elements', () => {
        // it('should to render correctly first field - label and input', () => {
        //     const formikWrapper = wrapper
        //                             .find(Formik)
        //                             .renderProp('children')(formikProps);
    
        //     expect(formikWrapper.childAt(0).getElement().type).toBe('label');
        //     expect(formikWrapper.childAt(0).prop('htmlFor')).toBe('beerName1');
        //     expect(formikWrapper.childAt(0).text()).toBe('Beer Name');

        //     expect(formikWrapper.childAt(1).find(Field)).toHaveLength(1);
        //     expect(formikWrapper.childAt(1).prop('className')).toBe('newBeerName');
        //     expect(formikWrapper.childAt(1).prop('name')).toBe('beerName');
            
        //     expect(formikWrapper.childAt(2).find(ErrorMessage)).toHaveLength(1);
        //     expect(formikWrapper.childAt(2).prop('name')).toBe('beerName');
        // });

        it('should to render correctly first field - label and input (Match Elements)', () => {
            const formikWrapper = wrapper
                                    .find(Formik)
                                    .renderProp('children')(formikProps);

            const beerName = {
                id: 'beerName1',
                label: 'Beer Name',
                className: 'newBeerName',
                name: 'beerName'
            };
            
            const beerNameLabelElementMock = <label htmlFor={beerName.id}>{beerName.label}</label>;
            const beerNameInputElementMock = <Field className={beerName.className} name={beerName.name}/>;

            console.log(formikWrapper.childAt(1).debug());
            
            expect(formikWrapper.childAt(0).matchesElement(beerNameLabelElementMock)).toBeTruthy();
            expect(formikWrapper.childAt(1).find(Field)).toHaveLength(1);
            expect(formikWrapper.childAt(1).matchesElement(beerNameInputElementMock)).toBeTruthy();

            /*
                Problemas: Não consegui de forma fácil simular o ErrorMessage utilizando matchElements.
            */
            //console.log(errorMessage(beerName.name));
            //console.log(formikWrapper.childAt(2).find(ErrorMessage).debug());
            expect(formikWrapper.childAt(2).find(ErrorMessage)).toHaveLength(1);
            //expect(formikWrapper.childAt(2).matchesElement(errorMessage(beerName.name))).toBeTruthy();
        });

        // it('', () => {
        //     const formikWrapper = wrapper
        //                             .find(Formik)
        //                             .renderProp('children')(formikProps);
            
        //     const optionsChildren = formikWrapper.childAt(4).find(Field).children();

            
        //     const beerTypeOptions = ['ale', 'lager', 'stout'];
        //     const optionsElementMock: Array<T> = [];
        //     const children = formikWrapper.childAt(4).find(Field).children();
            
        //     beerTypeOptions.forEach((option, index) => {
        //         console.log(index);
        //         optionsElementMock.push(<option value={option} hidden={true}>{option}</option>);
        //     });
        //     expect(children).toMatchSnapshot();
        // });

        // /* Problemas para utilizar o matchesElements em Elementos com filhos.
        // */
        // // it('should to render correctly second field - label, input and error message', () => {
        // //     const formikWrapper = wrapper
        // //                             .find(Formik)
        // //                             .renderProp('children')(formikProps);

        // //     const beerType = {
        // //         id: 'beerType1',
        // //         label: 'Beer Type',
        // //         className: 'newBeerType',
        // //         component: 'select',
        // //         default: '',
        // //         name: 'beerType'
        // //     };

        // //     const beerTypeOptions = ['ale', 'lager', 'stout'];
        // //     const optionsElementMock: Array<T> = [];
        // //     const children = formikWrapper.childAt(4).find(Field).children();

        // //     beerTypeOptions.forEach((option, index) => {
        // //         console.log(index);
        // //         optionsElementMock.push(<option value={option} hidden={true}>{option}</option>);
        // //     });

        // //     console.log(children.debug());
        // //     expect(children).toMatchInlineSnapshot(optionsElementMock);

        // //     const beerTypeLabelElementMock = <label htmlFor={beerType.id}>{beerType.label}</label>;
        // //     const beerTypeInputElementMock = <Field 
        // //                                             id={beerType.id}
        // //                                             className={beerType.className}
        // //                                             name={beerType.name}
        // //                                             component={beerType.component}
        // //                                             defaultValue={beerType.default} >lalala</Field>;
    
        // //     expect(formikWrapper.childAt(3).matchesElement(beerTypeLabelElementMock)).toBeTruthy();
        // //     console.log(formikWrapper.childAt(4).find(Field).getElement());
        // //     console.log(beerTypeInputElementMock);
        // //     expect(formikWrapper.childAt(4).find(Field).getElement()).toBe(beerTypeInputElementMock);
            
        // // });

        // it('should to render correctly third field - checkbox', () => {
        //     const formikWrapper = wrapper
        //                             .find(Formik)
        //                             .renderProp('children')(formikProps);

        //     const checkboxElementMock = <CheckboxFormikField
        //                                     className='newHasCorn'
        //                                     name='hasCorn'
        //                                     label='Has Corn' />;

        //     expect(formikWrapper.childAt(6).find(CheckboxFormikField)).toHaveLength(1);
        //     expect(formikWrapper.childAt(6).find(CheckboxFormikField).matchesElement(checkboxElementMock)).toBeTruthy();
        // });

        // it('should to render correctly fourth field - label, input and error message', () => {
        //     const formikWrapper = wrapper
        //                             .find(Formik)
        //                             .renderProp('children')(formikProps);
    
        //     expect(formikWrapper.childAt(7).getElement().type).toBe('label');
        //     expect(formikWrapper.childAt(7).prop('htmlFor')).toBe('ingredients1');
        //     expect(formikWrapper.childAt(7).text()).toBe('Ingredients');

        //     expect(formikWrapper.childAt(8).find(Field)).toHaveLength(1);
        //     expect(formikWrapper.childAt(8).prop('id')).toBe('ingredients1');
        //     expect(formikWrapper.childAt(8).prop('className')).toBe('newIngredients');
        //     expect(formikWrapper.childAt(8).prop('component')).toBe('textarea');
        //     expect(formikWrapper.childAt(8).prop('name')).toBe('ingredients');

        //     expect(formikWrapper.childAt(9).find(ErrorMessage)).toHaveLength(1);
        //     expect(formikWrapper.childAt(9).prop('name')).toBe('ingredients');
        // });


    });

    // it('', () => {});
});