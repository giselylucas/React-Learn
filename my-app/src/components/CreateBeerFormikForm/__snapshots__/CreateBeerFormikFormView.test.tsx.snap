// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBeerFormikFormView Render correctly order of elements  1`] = `
Array [
  <option
    hidden={true}
    value="ale"
  >
    ale
  </option>,
  <option
    hidden={true}
    value="lager"
  >
    lager
  </option>,
  <option
    hidden={true}
    value="stout"
  >
    stout
  </option>,
]
`;
